File: src/storage/filesystem.ts
import { mkdir, readFile, writeFile, readdir, access } from 'node:fs/promises';
import { join, resolve } from 'node:path';
import { stringify } from 'csv-stringify/sync';
import type { RunStorage } from './types.js';
import { GeoStorageError } from '../core/errors.js';

export class FileSystemStorage implements RunStorage {
  private baseDir: string;

  constructor(baseDir: string) {
    this.baseDir = resolve(baseDir);
  }

  getRunPath(runId: string): string {
    return join(this.baseDir, runId);
  }

  async createRunDir(runId: string): Promise<string> {
    const runDir = this.getRunPath(runId);
    await mkdir(runDir, { recursive: true });
    await mkdir(join(runDir, 'results', 'by-provider'), { recursive: true });
    await mkdir(join(runDir, 'analysis'), { recursive: true });
    await mkdir(join(runDir, 'reports', 'charts'), { recursive: true });
    return runDir;
  }

  async saveMetadata(runId: string, metadata: Record<string, unknown>): Promise<void> {
    await this.saveJSON(runId, 'metadata.json', metadata);
  }

  async loadMetadata(runId: string): Promise<Record<string, unknown>> {
    return this.loadJSON(runId, 'metadata.json');
  }

  async saveJSON(runId: string, filename: string, data: unknown): Promise<void> {
    const filePath = join(this.getRunPath(runId), filename);
    const dir = filePath.substring(0, filePath.lastIndexOf('/'));
    await mkdir(dir, { recursive: true });
    try {
      await writeFile(filePath, JSON.stringify(data, null, 2), 'utf-8');
    } catch (err) {
      throw new GeoStorageError(
        `Failed to save ${filename}: ${err instanceof Error ? err.message : String(err)}`,
      );
    }
  }

  async loadJSON<T>(runId: string, filename: string): Promise<T> {
    const filePath = join(this.getRunPath(runId), filename);
    try {
      const content = await readFile(filePath, 'utf-8');
      return JSON.parse(content) as T;
    } catch (err) {
      throw new GeoStorageError(
        `Failed to load ${filename}: ${err instanceof Error ? err.message : String(err)}`,
      );
    }
  }

  async saveCSV(runId: string, filename: string, data: Record<string, unknown>[]): Promise<void> {
    if (data.length === 0) return;

    const filePath = join(this.getRunPath(runId), filename);
    const dir = filePath.substring(0, filePath.lastIndexOf('/'));
    await mkdir(dir, { recursive: true });

    const columns = Object.keys(data[0]);
    const csv = stringify(data, { header: true, columns });

    try {
      await writeFile(filePath, csv, 'utf-8');
    } catch (err) {
      throw new GeoStorageError(
        `Failed to save CSV ${filename}: ${err instanceof Error ? err.message : String(err)}`,
      );
    }
  }

  async listRuns(): Promise<string[]> {
    try {
      const entries = await readdir(this.baseDir, { withFileTypes: true });
      return entries
        .filter((e) => e.isDirectory() && e.name.startsWith('run-'))
        .map((e) => e.name)
        .sort()
        .reverse();
    } catch {
      return [];
    }
  }

  async exists(runId: string, filename?: string): Promise<boolean> {
    const filePath = filename
      ? join(this.getRunPath(runId), filename)
      : this.getRunPath(runId);
    try {
      await access(filePath);
      return true;
    } catch {
      return false;
    }
  }
}

File: src/voyage_geo/storage/filesystem.py
"""File-based storage for run data."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any

import structlog

logger = structlog.get_logger()


class FileSystemStorage:
    def __init__(self, base_dir: str = "./data/runs") -> None:
        self.base_dir = Path(base_dir)

    async def create_run_dir(self, run_id: str) -> Path:
        run_dir = self.base_dir / run_id
        for sub in [
            "",
            "results",
            "results/by-provider",
            "analysis",
            "reports",
            "reports/charts",
        ]:
            (run_dir / sub).mkdir(parents=True, exist_ok=True)
        return run_dir

    def run_dir(self, run_id: str) -> Path:
        return self.base_dir / run_id

    async def save_json(self, run_id: str, filename: str, data: Any) -> Path:
        path = self.base_dir / run_id / filename
        path.parent.mkdir(parents=True, exist_ok=True)

        # Convert Pydantic models to dicts
        if hasattr(data, "model_dump"):
            data = data.model_dump()

        with open(path, "w") as f:
            json.dump(data, f, indent=2, default=str)

        logger.debug("storage.saved", path=str(path))
        return path

    async def load_json(self, run_id: str, filename: str) -> Any:
        path = self.base_dir / run_id / filename
        if not path.exists():
            return None
        with open(path) as f:
            return json.load(f)

    async def save_metadata(self, run_id: str, metadata: dict) -> None:
        await self.save_json(run_id, "metadata.json", metadata)

    async def save_text(self, run_id: str, filename: str, text: str) -> Path:
        path = self.base_dir / run_id / filename
        path.parent.mkdir(parents=True, exist_ok=True)
        with open(path, "w") as f:
            f.write(text)
        return path

    def list_runs(self) -> list[str]:
        if not self.base_dir.exists():
            return []
        return sorted(
            [d.name for d in self.base_dir.iterdir() if d.is_dir() and (d.name.startswith("run-") or d.name.startswith("lb-"))],
            reverse=True,
        )
